File: src/types/Tower.go
package types

import (
	"fmt"
	"regexp"
	"strconv"
	"time"

	"Jazzmoon/btd6_autoplay/utils"

	"github.com/go-vgo/robotgo"
	"gopkg.in/yaml.v3"
)

type Tower struct {
	Name   string   `yaml:"name,omitempty"`
	Hotkey []string `yaml:"hotkey"`
	Coords Coords   `yaml:"coords"`
	Path   [3]int
}

/*
String is a function that returns a string representation of the Tower object
  - @return string: A string that contains the name, hotkey, coordinates, and path of the tower
*/
func (t *Tower) String() string {
	return fmt.Sprintf("Name: %s, Hotkey: %+v, Coords: %v, Path: %v", t.Name, t.Hotkey, t.Coords, t.Path)
}

/*
InitTower is a function that initializes a Tower object with the given name, hotkey, and coordinates
  - @param name: A string that indicates the name of the tower
  - @param hotkey: A string that indicates the hotkey of the tower
  - @param x: An integer that indicates the x-coordinate of the tower
  - @param y: An integer that indicates the y-coordinate of the tower
  - @return Tower: A Tower object that contains the name, hotkey, and coordinates of the tower
*/
func InitTower(name string, hotkey []string, x, y int) Tower {
	return Tower{
		Name:   name,
		Hotkey: hotkey,
		Coords: Coords{
			X: x,
			Y: y,
		},
		Path: [3]int{0, 0, 0},
	}
}

/*
UnmarshalYAML is a function that allows the user to unmarshal a YAML object into a Tower object
  - @param unmarshal: A function that unmarshals the YAML object into a Tower object
  - @return interface{}: An interface that contains the Tower object
  - @return error: An error that indicates why the Tower object was not unmarshaled successfully
*/
func (t *Tower) UnmarshalYAML(value *yaml.Node) error {
	var tower struct {
		Name   string   `yaml:"name"`
		Hotkey []string `yaml:"hotkey"`
		Coords Coords   `yaml:"coords"`
	}
	if err := value.Decode(&tower); err != nil {
		return err
	}
	t.Name = tower.Name
	t.Hotkey = tower.Hotkey
	t.Coords = tower.Coords
	t.Path = [3]int{0, 0, 0}
	return nil
}

/*
Deselect is a function that allows the user to deselect a tower on the screen
  - @return err: An error that indicates why the tower was not deselected successfully
*/
func (t *Tower) Deselect() error {
	// Move to center of screen and click to deselect the tower
	utils.SleepMoveAndClick(2, CurrentMap.HoverLocation.X, CurrentMap.HoverLocation.Y, "left")
	return nil
}

/*
Place is a function that allows the user to place a tower on the screen
  - @return err: An error that indicates why the tower was not placed successfully
*/
func (t *Tower) Place() error {
	mouseSleep, keySleep := robotgo.MouseSleep, robotgo.KeySleep
	robotgo.MouseSleep, robotgo.KeySleep = 2, 2
	robotgo.Move(t.Coords.X, t.Coords.Y)
	if err := robotgo.KeyTap(t.Hotkey[0], t.Hotkey[1:]); err != nil {
		return err
	}
	robotgo.Click("left")
	if err := t.Deselect(); err != nil {
		return err
	}
	robotgo.MouseSleep, robotgo.KeySleep = mouseSleep, keySleep
	return nil
}

/*
Highlight is a function that allows the user to highlight a tower on the screen
  - @param args: An optional boolean that indicates whether to select the tower or not
  - @return err: An error that indicates why the tower was not highlighted successfully
*/
func (t *Tower) Highlight(args ...bool) error {
	if len(args) > 1 {
		return fmt.Errorf("[Tower | %s] Invalid number of arguments. Expected 0 or 1, got %d", t.Name, len(args))
	}
	if len(args) == 1 && args[0] {
		utils.SleepMoveAndClick(2, t.Coords.X, t.Coords.Y, "left")
	} else {
		utils.SleepMove(2, t.Coords.X, t.Coords.Y)
	}
	return nil
}

/*
Upgrade is a function that allows the user to upgrade a tower on the screen
  - @param path: A string that indicates the path of the upgrade to select in the format of `\d-\d-\d` (e.g. `0-0-0`)
  - @return err: An error that indicates why the tower was not upgraded successfully
*/
func (t *Tower) Upgrade(path string) error {
	// First, we parse the upgrade path string, in the format of `\d-\d-\d` (e.g. `0-0-0`) into an array of integers
	desiredPath := [3]int{0, 0, 0}
	upgradePathRE, err := regexp.Compile(`(\d)-(\d)-(\d)`)
	if err != nil {
		return err
	}
	upgradePath := upgradePathRE.FindStringSubmatch(path)
	if upgradePath == nil {
		return fmt.Errorf("[Tower | %s] Failed to match expected path pattern", t.Name)
	}
	// Remove the first element of upgradePath, which is the full match
	upgradePath = upgradePath[1:]

	for i, path := range upgradePath {

		desiredPath[i], err = strconv.Atoi(path)
		if err != nil {
			return err
		}
	}
	// Next, we calculate the difference between the current path and the desired path to determine which upgrade(s) to select
	diff := [3]int{desiredPath[0] - t.Path[0], desiredPath[1] - t.Path[1], desiredPath[2] - t.Path[2]}
	// Finally, we select the upgrade(s) by pressing the corresponding hotkey(s)
	if err := t.Highlight(true); err != nil {
		return err
	}
	time.Sleep(50 * time.Millisecond)

	keySleep := robotgo.KeySleep
	robotgo.KeySleep = 50
	if diff[0] > 0 {
		for i := 0; i < diff[0]; i++ {
			if err := robotgo.KeyTap(Settings.Game.Hotkeys.UpgradeTopPath[0], Settings.Game.Hotkeys.UpgradeTopPath[1:]); err != nil {
				return err
			}
		}
	}
	if diff[1] > 0 {
		for i := 0; i < diff[1]; i++ {
			if err := robotgo.KeyTap(Settings.Game.Hotkeys.UpgradeMiddlePath[0], Settings.Game.Hotkeys.UpgradeMiddlePath[1:]); err != nil {
				return err
			}
		}
	}
	if diff[2] > 0 {
		for i := 0; i < diff[2]; i++ {
			if err := robotgo.KeyTap(Settings.Game.Hotkeys.UpgradeBottomPath[0], Settings.Game.Hotkeys.UpgradeBottomPath[1:]); err != nil {
				return err
			}
		}
	}
	robotgo.KeySleep = keySleep
	// Update the tower's path to reflect the upgrade
	t.Path = desiredPath

	utils.SleepMoveAndClick(1, CurrentMap.HoverLocation.X, CurrentMap.HoverLocation.Y, "left")

	return nil
}

/*
Sell is a function that allows the user to sell a tower on the screen
  - @return err: An error that indicates why the tower was not sold successfully
*/
func (t *Tower) Sell() error {
	if err := t.Highlight(true); err != nil {
		return err
	}
	if err := utils.SleepKeyTap(2, Settings.Game.Hotkeys.Sell[0], Settings.Game.Hotkeys.Sell[1:]); err != nil {
		return err
	}
	if err := t.Deselect(); err != nil {
		return err
	}
	return nil
}

/*
GetPath is a function that returns the path of the tower
  - @return [3]int: An array of integers that contains the path of the tower
*/
func (t *Tower) GetPath() [3]int {
	return t.Path
}

/*
Reset is a function that resets the path of the tower for game restarts
*/
func (t *Tower) Reset() {
	t.Path = [3]int{0, 0, 0}
}

File: src/models/tower.rs
use super::{coords::Coords, hotkeys::Hotkey};
use crate::utils::{
    global::{CURRENT_MAP, HOTKEYS},
    interaction,
};

use enigo::InputResult;
use serde::{Deserialize, Serialize};



#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct Tower {
    pub name: String,
    pub hotkey: Hotkey,
    pub coords: Coords,
    pub upgrade_path: [i32; 3],
}

impl Tower {
    pub fn new(name: String, hotkey: Hotkey, coords: Coords, path: Option<[i32; 3]>) -> Tower {
        Tower {
            name,
            hotkey,
            coords,
            upgrade_path: path.unwrap_or([0, 0, 0]),
        }
    }

    pub fn to_string(&self) -> String {
        format!(
            "Tower: {}, Hotkey: {:?}, Coords: {:?}, Path: {:?}",
            self.name, self.hotkey, self.coords, self.upgrade_path
        )
    }

    pub fn highlight(&self, perform_click: Option<bool>) -> InputResult<()> {
        let click = perform_click.unwrap_or(false);
        if click {
            interaction::click(self.coords.clone(), None)
        } else {
            interaction::move_cursor(self.coords.clone(), None)
        }
    }

    pub fn deselect(&self) -> InputResult<()> {
        let current_map_read_lock = CURRENT_MAP.read().unwrap();
        let current_map = current_map_read_lock.as_ref().unwrap();
        let hover_location = current_map.hover_location.clone();
        interaction::click(hover_location, None)
    }

    pub fn place(&self) -> Result<(), String> {
        let _ = interaction::press_key(self.hotkey.clone(), None);
        let _ = interaction::click(self.coords.clone(), None);
        match self.deselect() {
            Ok(_) => Ok(()),
            Err(e) => Err(e.to_string()),
        }
    }

    pub fn upgrade(&mut self, upgrade_path: String) -> Result<[i32; 3], String> {
        // First, parse the upgrade path string (`\d-\d-\d`) into an array of 3 integers
        let upgrade_path: Vec<i32> = upgrade_path
            .split('-')
            .map(|x| x.parse::<i32>().unwrap())
            .collect();

        if upgrade_path.len() != 3 {
            return Err(format!(
                "Invalid upgrade path: {:?} for tower: {:?}",
                upgrade_path, self.name
            ));
        }

        // Next, calculate the difference vector between the current upgrade path and the new upgrade path
        let diff: [i32; 3] = upgrade_path
            .iter()
            .zip(self.upgrade_path.iter())
            .map(|(x, y)| x - y)
            .collect::<Vec<i32>>()
            .try_into()
            .unwrap();

        // Check to make sure the upgrade path is valid (we cannot downgrade a tower)
        if !diff.iter().all(|&x| x >= 0) {
            // If the upgrade path is invalid, print an error message
            return Err(format!(
                "Invalid upgrade path: {:?} for tower: {:?}",
                upgrade_path, self.name
            ));
        }
        // Finally, update the tower's upgrade path by Selecting it, and pressing the upgrade hotkeys in order
        let res = self.highlight(Some(true));
        if res.is_err() {
            return Err(format!("Failed to select tower: {:?}", self.name));
        }

        {
            let hotkeys_read_lock = HOTKEYS.read().unwrap();
            let hotkeys = hotkeys_read_lock.as_ref().unwrap();

            if diff[0] > 0 {
                for _ in 0..diff[0] {
                    let _ = interaction::press_key(hotkeys.upgrade_top_path.clone(), None);
                }
            }

            if diff[1] > 0 {
                for _ in 0..diff[1] {
                    let _ = interaction::press_key(hotkeys.upgrade_middle_path.clone(), None);
                }
            }

            if diff[2] > 0 {
                for _ in 0..diff[2] {
                    let _ = interaction::press_key(hotkeys.upgrade_bottom_path.clone(), None);
                }
            }
        }

        self.upgrade_path = upgrade_path.try_into().unwrap();
        match self.deselect() {
            Ok(_) => Ok(self.upgrade_path),
            Err(e) => Err(e.to_string()),
        }
    }

    pub fn sell(&self) -> InputResult<()> {
        let res = self.highlight(Some(true));
        if res.is_err() {
            return res;
        }
        let hotkeys_read_lock = HOTKEYS.read().unwrap();
        let hotkeys = hotkeys_read_lock.as_ref().unwrap();
        let _ = interaction::press_key(hotkeys.sell.clone(), None);
        Ok(())
    }
}
